### PR TITLE
[Merged by Bors] - combine remainders of ET-3845 and non nc back ported parts

### PR DESCRIPTION
**Pull request by rustonaut**
- To avoid messy semi-circular dependency circular a new `integration-test` crate is introduced which under `lib.rs` has tooling for integration tests and places the tests under `tests`.
 
 - `AppHandle` is extended to have a `on_shutdown` callback, which is used to close the postgres storage pool or the db deletion trick won't work
 
 - env cleaning and bindings to the test setup/teardown parts in the justfile are taken from the remainders of ET-3845
 
 - `test_app` is taken from the nc repo and combined with a similar helper from the remainders or ET-3845
 
 - a test for the inline config from ET-3845 was added
 
 - `Panic` error type to allow using `?` in tests while still getting nice backtraces and early exit 
 
 **References:**

- private nc repo

- ET-3845, #762

### CODE REVIEW

**Comment by rustonaut** [comment]
The yellow image is currently missing the `psql` command.

**Comment by rustonaut** [comment]
The reason why this uses `just` actions to setup/teardown tests services was so that it can also be used in python.

Through how it currently looks we will end up more with:

- detailed well isolated integration tests in rust, but the tests are coupled to the local test setup
- more generic not isolated integration tests in python which we could point to a dev deployment

This made me think about two things:

- :wrench: (?) should we keep db setup/teardown in just and add psql to yellow docker image or should I checkout older code from ET-3845 where this parts where in rust (using reqwest for ES and sqlx fro PG)
- :pushpin: (?) maybe python wasn't the best choice, putting rust aside wouldn't JS/TS on node using the generate JS API  have been a better choice?

File: integration-tests/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+    let output = String::from_utf8(stdout)?;
+    if status.success() {
+        Ok(output)
+    } else {
+        bail!("Cmd `just` failed! Output:\n{output}Just Exit Status: {status}");
+    }
+}
+
+pub async fn send_assert_success(client: &Client, req: Request) {
```
Comment (rustonaut) [inline integration-tests/src/lib.rs:68]: I guess this should be changed to allow checking for specific status codes.

File: integration-tests/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+    let storage = config["storage"].as_table_mut().unwrap();
+    storage["postgres"]
+        .as_table_mut()
+        .unwrap()
+        .insert("base_url".into(), services.postgres.clone().into());
+    let elastic = storage["elastic"].as_table_mut().unwrap();
+    let (url, index) = services.elastic_search.rsplit_once('/').unwrap();
+    elastic.insert("url".into(), url.into());
+    elastic.insert("index".into(), index.into());
```
Comment (rustonaut) [inline integration-tests/src/lib.rs:132]: I'm not the biggest fan how this turned out wrt. readability, but for now that is good enough I guess.

**Review by janpetschexain — COMMENTED** [review]
> 🔧 (?) should we keep db setup/teardown in just and add psql to yellow docker image or should I checkout older code from ET-3845 where this parts where in rust (using reqwest for ES and sqlx fro PG)

i'll update the image to include `psql` in a separate task

File: test-utils/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+use std::{convert::Infallible, fmt::Write};
+
+/// Error which panics when creating.
+///
+/// Useful for using `?` in tests which will panic on error anyway.
+#[derive(Debug)]
+pub struct Panic {
+    _priv: Infallible,
+}
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply test-utils/src/error.rs:23]: Reasoning: `Panic` can never be constructed hence why `Panic { _priv: ! }` would be the right thing, and the next closest thing to `!` is `Infallible`.

But I agree that it reads quite confusing so I will change it.

File: integration-tests/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+        let bytes = response.bytes().await.unwrap();
+        let text = String::from_utf8_lossy(&bytes);
+        panic!("Failed to {method} {target}, status {status} instead of {expected}. Body:\n{text}");
+    }
+    response
+}
+
+pub async fn send_assert_json<O>(client: &Client, req: Request, expected: StatusCode) -> O
+where
+    O: DeserializeOwned,
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply integration-tests/src/lib.rs:83]: This is the exact definition of `DeserializeOwned`.

```rust
pub trait DeserializeOwned: for<'de> Deserialize<'de> {}
impl<T> DeserializeOwned for T where T: for<'de> Deserialize<'de> {}
```

In the end we return a owned value which can't internally reference the buffer it was deserialized from, so I think `DeserializeOwned` is more clear.

File: integration-tests/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+}
+
+#[derive(Clone, Debug)]
+pub struct Services {
+    /// Id of the test.
+    pub id: String,
+    /// Uri to a postgres db for this test.
+    pub postgres: String,
+    /// Uri to a elastic search db for this test.
+    pub elastic_search: String,
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply integration-tests/src/lib.rs:275]: we could but it doesn't add much value here and in some older versions it caused more problems then it helped

File: integration-tests/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (janpetschexain) [reply integration-tests/src/lib.rs:83]: i agree, i didn't know that `DeserializeOwned` is just a helper trait for the more verbose one 😅

**Review by janpetschexain — APPROVED** [review]
lgtm 👍 

❓ is there a way to reduce the verbosity of the just scripts a bit? it produces quite some noise now if you run `just test`.

**Comment by rustonaut** [comment]
> question is there a way to reduce the verbosity of the just scripts a bit? it produces quite some noise now if you run just test.

Something of our app brakes the custom output capturing, probably tracing, probably setting a panic hook.

So we should be able to do something about it in a follow up PR:

1. an config option to not set the panic hook
2. create a temp dir additionally to the pg db/es index and pipe server log messages there and not to stdout at all
   - (also clean it up when the test succeeds)

File: test-utils/src/error.rs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+use std::fmt::Write;
+
+/// Error which panics when creating.
+///
+/// Useful for using `?` in tests which will panic on error anyway.
+#[derive(Debug)]
+pub struct Panic {
+    _priv: (),
+}
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply test-utils/src/error.rs]: good idea :rocket:

**Comment by rustonaut** [comment]
I also removed the `-x` bash option from `just` command called many times , that already removed a lot of verbosity.

**Pull request closed without merging**